core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -106,23 +96,81 @@ public Page getPage()
         return page;
     }
 
-    private boolean currentRowContainsNull()
+    private static long[] fillCache(
+            LookupSource lookupSource,
+            Page page,
+            Block probeHashBlock,
+            Page probePage)
     {
-        for (int i = 0; i < probePage.getChannelCount(); i++) {
-            if (probePage.getBlock(i).isNull(position)) {
-                return true;
+        int positionCount = page.getPositionCount();
+        List<Block> nullableBlocks = IntStream.range(0, probePage.getChannelCount())
+                .mapToObj(i -> probePage.getBlock(i))
+                .filter(Block::mayHaveNull)
+                .collect(toImmutableList());
+
+        long[] joinPositionCache = new long[positionCount];
+        if (!nullableBlocks.isEmpty()) {
+            Arrays.fill(joinPositionCache, -1);
+            boolean[] isNull = new boolean[positionCount];
+            int nonNullCount = getIsNull(nullableBlocks, positionCount, isNull);
+            if (nonNullCount < positionCount) {
+                // We only store positions that are not null
+                int[] positions = new int[nonNullCount];
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        positions[nonNullCount] = i;
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                if (probeHashBlock != null) {
+                    long[] hashes = new long[positionCount];
+                    for (int i = 0; i < positionCount; i++) {
+                        hashes[i] = BIGINT.getLong(probeHashBlock, i);
+                    }
+                    lookupSource.getJoinPosition(positions, probePage, page, hashes, joinPositionCache);
+                }
+                else {
+                    lookupSource.getJoinPosition(positions, probePage, page, joinPositionCache);
+                }
+                return joinPositionCache;
```

Comment:
Is it possible to avoid duplicating this block of code  in the else branch ? The only difference appears to be in the contents of `positions`.

Reply:
The first one has additional `hashes` argument